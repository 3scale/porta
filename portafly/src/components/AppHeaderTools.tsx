import React, { useState } from 'react'
import { useAuth } from 'auth'
import {
  Button,
  PageHeaderTools,
  PageHeaderToolsGroup,
  PageHeaderToolsItem,
  ButtonVariant,
  Dropdown,
  DropdownToggle,
  Avatar,
  KebabToggle,
  DropdownItem,
  DropdownGroup
} from '@patternfly/react-core'
import { useTranslation } from 'i18n/useTranslation'
import { CogIcon, HelpIcon } from '@patternfly/react-icons'
import avatar from 'assets/img_avatar.svg'

const AppHeaderTools: React.FunctionComponent = () => {
  const { t } = useTranslation('shared')
  const { userProfile, logout } = useAuth()

  const [isKebabDropdownOpen, setIsKebabDropdownOpen] = useState(false)
  const [isDropdownOpen, setIsDropDownOpen] = useState(false)

  const kebabDropdownItems = [
    <DropdownItem key="settings">
      <CogIcon />
      {' Settings'}
    </DropdownItem>,
    <DropdownItem key="help">
      <HelpIcon />
      {' Help'}
    </DropdownItem>
  ]

  const userDropdownItems = [
    <DropdownGroup key="group user">
      <DropdownItem key="group user logout" onClick={() => logout()}>
        {t('buttons.logout')}
      </DropdownItem>
    </DropdownGroup>
  ]

  return (
    <PageHeaderTools>
      <PageHeaderToolsGroup visibility={{ default: 'hidden', lg: 'visible' }}>
        <PageHeaderToolsItem>
          <Button aria-label="Settings actions" variant={ButtonVariant.plain}>
            <CogIcon />
          </Button>
        </PageHeaderToolsItem>
        <PageHeaderToolsItem>
          <Button aria-label="Help actions" variant={ButtonVariant.plain}>
            <HelpIcon />
          </Button>
        </PageHeaderToolsItem>
      </PageHeaderToolsGroup>
      <PageHeaderToolsGroup>
        <PageHeaderToolsItem visibility={{
          default: 'visible',
          sm: 'visible',
          md: 'visible',
          lg: 'hidden',
          xl: 'hidden',
          '2xl': 'hidden'
        }}
        >
          <Dropdown
            isPlain
            position="right"
            onSelect={() => setIsKebabDropdownOpen(!isKebabDropdownOpen)}
            toggle={<KebabToggle onToggle={() => setIsKebabDropdownOpen(!isKebabDropdownOpen)} />}
            isOpen={isKebabDropdownOpen}
            dropdownItems={kebabDropdownItems}
          />
        </PageHeaderToolsItem>
        <PageHeaderToolsItem visibility={{
          default: 'hidden',
          sm: 'hidden',
          md: 'hidden',
          lg: 'visible',
          xl: 'visible',
          '2xl': 'visible'
        }}
        >
          <Dropdown
            isPlain
            position="right"
            onSelect={() => setIsDropDownOpen(!isDropdownOpen)}
            isOpen={isDropdownOpen}
            toggle={(
              <DropdownToggle onToggle={() => setIsDropDownOpen(!isDropdownOpen)}>
                {`${userProfile.firstName} ${userProfile.lastName}`}
              </DropdownToggle>
              )}
            dropdownItems={userDropdownItems}
          />
        </PageHeaderToolsItem>
      </PageHeaderToolsGroup>
      <Avatar src={avatar} alt="Avatar image" />
    </PageHeaderTools>
  )
}

export { AppHeaderTools }
