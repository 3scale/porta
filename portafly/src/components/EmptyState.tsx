import React from 'react'
import {
  Title,
  EmptyState,
  EmptyStateVariant,
  EmptyStateIcon,
  Bullseye,
  EmptyStateBody
} from '@patternfly/react-core'
import { CubesIcon, SearchIcon } from '@patternfly/react-icons'

export interface ISimpleEmptyStateProps {
  msg: string
}

const SimpleEmptyState: React.FunctionComponent<ISimpleEmptyStateProps> = ({ msg }) => (
  <EmptyState variant={EmptyStateVariant.full}>
    <EmptyStateIcon icon={CubesIcon} />
    <Title headingLevel="h5" size="lg">
      {msg}
    </Title>
  </EmptyState>
)

interface ITableEmptyState {
  title: string
  body: string
  button?: JSX.Element
}

const TableEmptyState: React.FunctionComponent<ITableEmptyState> = ({
  title,
  body,
  button
}) => (
  <Bullseye>
    <EmptyState variant={EmptyStateVariant.small}>
      <EmptyStateIcon icon={SearchIcon} />
      <Title headingLevel="h2" size="lg">
        {title}
      </Title>
      <EmptyStateBody>
        {body}
      </EmptyStateBody>
      {button}
    </EmptyState>
  </Bullseye>
)

export { SimpleEmptyState, TableEmptyState }
