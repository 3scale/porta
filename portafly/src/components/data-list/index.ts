export * from './BulkSelectorWidget'
export * from './BulkActionsWidget'
export * from './DataListContext'
export * from './SearchWidget'
export * from './Toolbar'
export * from './PaginationWidget'
export * from './modals'
export * from './utils'
