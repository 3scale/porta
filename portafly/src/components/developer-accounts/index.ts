export * from './ActionsDropdown'
export * from './BulkActionSelector'
export * from './SearchWidget'
export * from './DeveloperAccountsTable'
export * from './ActionButtonImpersonate'
export * from './util/tableDataFactory'
