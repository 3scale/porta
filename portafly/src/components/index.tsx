export * from './pages'
export * from './shared'
export * from './AppLayout'
export * from './AppNavExpandable'
export * from './AppNavGroup'
export * from './AppNavItem'
export * from './AppHeaderTools'
export * from './LazyRoute'
export * from './NotFound'
export * from './SwitchWith404'
export * from './util'
export * from './Root'
