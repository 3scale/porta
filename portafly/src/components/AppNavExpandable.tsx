import * as React from 'react'
import { useRouteMatch } from 'react-router-dom'
import { NavExpandable } from '@patternfly/react-core'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppNavExpandableProps {
  title: string
  to: string
  items: Array<IAppNavItemProps | undefined>
}

export const AppNavExpandable: React.FunctionComponent<IAppNavExpandableProps> = ({
  title,
  to,
  items
}) => {
  const match = useRouteMatch({
    path: to
  })
  const isActive = !!match
  return (
    <NavExpandable title={title} isActive={isActive} isExpanded={isActive}>
      {items.map((item, j) => (
        // FIXME: each app-nav-item should have a fixed ID, using title? or j is just a dirty trick
        <AppNavItem title={item?.title} to={item?.to} key={item?.title || j} />
      ))}
    </NavExpandable>
  )
}
