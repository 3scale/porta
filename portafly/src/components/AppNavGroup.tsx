import * as React from 'react'
import { NavGroup } from '@patternfly/react-core'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppNavGroupProps {
  title: string
  items: Array<IAppNavItemProps | undefined>
}

export const AppNavGroup: React.FunctionComponent<IAppNavGroupProps> = ({
  title,
  items
}) => (
  <NavGroup title={title}>
    {items.map((i) => (
      <AppNavItem title={i?.title} to={i?.to} exact={i?.exact} key={i?.title} />
    ))}
  </NavGroup>
)
