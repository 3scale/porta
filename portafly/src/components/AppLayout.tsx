import * as React from 'react'
import {
  Nav,
  NavList,
  Page,
  PageHeader,
  PageSidebar,
  SkipToContent,
  PageHeaderProps
} from '@patternfly/react-core'
import { AppNavExpandable, IAppNavExpandableProps } from 'components/AppNavExpandable'
import { AppNavGroup, IAppNavGroupProps } from 'components/AppNavGroup'
import { AppNavItem, IAppNavItemProps } from 'components/AppNavItem'

export interface IAppLayoutContext {
  setBreadcrumb: (breadcrumb: React.ReactNode) => void
}

export const AppLayoutContext = React.createContext<IAppLayoutContext>({
  setBreadcrumb: () => 0
})

export interface IAppLayoutProps extends Pick<PageHeaderProps, 'logo' | 'logoProps' | 'headerTools'> {
  navVariant?: 'vertical' | 'horizontal'
  navItems?: Array<IAppNavItemProps | IAppNavExpandableProps | IAppNavGroupProps | undefined>
  navGroupsStyle?: 'grouped' | 'expandable'
  startWithOpenNav?: boolean
  theme?: 'dark' | 'light'
  mainContainerId?: string
}

export const AppLayout: React.FunctionComponent<IAppLayoutProps> = ({
  logo,
  logoProps,
  navVariant = 'horizontal',
  navItems = [],
  navGroupsStyle = 'grouped',
  headerTools,
  startWithOpenNav = true,
  theme = 'dark',
  mainContainerId = 'main-container',
  children
}) => {
  const [isNavOpen, setIsNavOpen] = React.useState(startWithOpenNav)
  const [isMobileView, setIsMobileView] = React.useState(true)
  const [isNavOpenMobile, setIsNavOpenMobile] = React.useState(false)
  const [breadcrumb, setBreadcrumb] = React.useState<React.ReactNode | undefined>()
  const previousBreadcrumb = React.useRef<React.ReactNode | null>()

  const handleSetBreadcrumb = React.useCallback(
    (newBreadcrumb: React.ReactNode) => {
      if (previousBreadcrumb.current !== newBreadcrumb) {
        previousBreadcrumb.current = newBreadcrumb
        setBreadcrumb(previousBreadcrumb.current)
      }
    },
    [setBreadcrumb, previousBreadcrumb]
  )

  const onNavToggleMobile = React.useCallback(() => {
    setIsNavOpenMobile(!isNavOpenMobile)
  }, [setIsNavOpenMobile, isNavOpenMobile])

  const onNavToggle = React.useCallback(() => {
    setIsNavOpen(!isNavOpen)
  }, [setIsNavOpen, isNavOpen])

  const onPageResize = (props: { mobileView: boolean; windowSize: number }) => {
    setIsMobileView(props.mobileView)
  }

  React.useEffect(() => {
    setIsNavOpen(startWithOpenNav)
  }, [startWithOpenNav, setIsNavOpen])

  const isVertical = navVariant === 'vertical'
  const variant = isVertical ? 'default' : 'horizontal'

  const Navigation = React.useMemo(
    () => navItems.length > 0 && (
    <Nav id="nav-primary-simple" theme={theme} variant={variant}>
      <NavList id="nav-list-simple">
        {navItems.map((navItem) => {
          // eslint-disable-next-line no-prototype-builtins
          if (navItem && navItem.hasOwnProperty('items') && isVertical) {
            if (navGroupsStyle === 'expandable') {
              const { title, to, items } = navItem as IAppNavExpandableProps
              return <AppNavExpandable title={title} to={to} items={items} key={title} />
            }

            const { title, items } = navItem as IAppNavGroupProps
            return <AppNavGroup title={title} items={items} key={title} />
          }

          // TODO: extract NavItemSeparator from AppNavItem to clean this up
          const item = navItem as (IAppNavItemProps | undefined)
          return (
            <AppNavItem
              title={item?.title}
              to={item?.to}
              exact={item?.exact}
              key={item?.title}
            />
          )
        })}
      </NavList>
    </Nav>
    ),
    [isVertical, navGroupsStyle, navItems, theme, variant]
  )

  const Header = React.useMemo(
    () => (
      <PageHeader
        logo={logo}
        logoProps={logoProps}
        headerTools={headerTools}
        showNavToggle={isVertical}
        isNavOpen={isVertical && isNavOpen}
        onNavToggle={isMobileView ? onNavToggleMobile : onNavToggle}
        topNav={isVertical ? undefined : Navigation}
      />
    ),
    [
      logo,
      logoProps,
      headerTools,
      isVertical,
      isNavOpen,
      isMobileView,
      onNavToggle,
      onNavToggleMobile,
      Navigation
    ]
  )

  const Sidebar = React.useMemo(
    () => navVariant === 'vertical' && (
    <PageSidebar
      nav={Navigation}
      isNavOpen={isMobileView ? isNavOpenMobile : isNavOpen}
      theme={theme}
      data-testid="app-sidebar"
    />
    ),
    [navVariant, Navigation, isMobileView, isNavOpenMobile, isNavOpen, theme]
  )
  const PageSkipToContent = (
    <SkipToContent href="#primary-app-container">
      Skip to Content
    </SkipToContent>
  )
  return (
    <AppLayoutContext.Provider value={{ setBreadcrumb: handleSetBreadcrumb }}>
      <Page
        mainContainerId={mainContainerId}
        header={Header}
        sidebar={Sidebar}
        breadcrumb={breadcrumb}
        onPageResize={onPageResize}
        skipToContent={PageSkipToContent}
      >
        {children}
      </Page>
    </AppLayoutContext.Provider>
  )
}
