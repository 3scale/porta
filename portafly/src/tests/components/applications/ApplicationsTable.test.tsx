import * as React from 'react'

import { render } from 'tests/setup'
import { ApplicationsTable } from 'components'

export const applications = [{
  name: 'Application Name',
  state: 'Some status',
  account: 'Developer',
  plan: { name: 'Plan Name' },
  created_at: 1234567890
}]

describe('when there are any applications', () => {

  it('should render a table with applications', () => {
    const wrapper = render(<ApplicationsTable applications={applications} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('when there are NO applications', () => {

  it('should render an empty view', () => {
    const wrapper = render(<ApplicationsTable applications={[]} />)
    expect(wrapper).toMatchSnapshot()
  })
})
