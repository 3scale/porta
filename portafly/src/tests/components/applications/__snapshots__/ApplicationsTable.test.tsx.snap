// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when there are NO applications should render an empty view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-c-empty-state"
      >
        <svg
          aria-hidden="true"
          class="pf-c-empty-state__icon"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
            transform=""
          />
        </svg>
        <h5
          class="pf-c-title pf-m-lg"
        >
          There are no Applications
        </h5>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-c-empty-state"
    >
      <svg
        aria-hidden="true"
        class="pf-c-empty-state__icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
          transform=""
        />
      </svg>
      <h5
        class="pf-c-title pf-m-lg"
      >
        There are no Applications
      </h5>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`when there are any applications should render a table with applications 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        aria-label="Applications list"
        class="pf-c-table pf-m-grid-md"
        role="grid"
      >
        <thead
          class=""
        >
          <tr>
            <th
              class=""
              data-key="0"
              data-label="Name"
              scope="col"
            >
              Name
            </th>
            <th
              class=""
              data-key="1"
              data-label="State"
              scope="col"
            >
              State
            </th>
            <th
              class=""
              data-key="2"
              data-label="Account"
              scope="col"
            >
              Account
            </th>
            <th
              class=""
              data-key="3"
              data-label="Plan"
              scope="col"
            >
              Plan
            </th>
            <th
              class=""
              data-key="4"
              data-label="Created at"
              scope="col"
            >
              Created at
            </th>
          </tr>
        </thead>
        <tbody
          class=""
        >
          <tr
            class=""
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              Application Name
            </td>
            <td
              class=""
              data-key="1"
              data-label="State"
            >
              Some status
            </td>
            <td
              class=""
              data-key="2"
              data-label="Account"
            >
              Developer
            </td>
            <td
              class=""
              data-key="3"
              data-label="Plan"
            >
              Plan Name
            </td>
            <td
              class=""
              data-key="4"
              data-label="Created at"
            >
              1234567890
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      aria-label="Applications list"
      class="pf-c-table pf-m-grid-md"
      role="grid"
    >
      <thead
        class=""
      >
        <tr>
          <th
            class=""
            data-key="0"
            data-label="Name"
            scope="col"
          >
            Name
          </th>
          <th
            class=""
            data-key="1"
            data-label="State"
            scope="col"
          >
            State
          </th>
          <th
            class=""
            data-key="2"
            data-label="Account"
            scope="col"
          >
            Account
          </th>
          <th
            class=""
            data-key="3"
            data-label="Plan"
            scope="col"
          >
            Plan
          </th>
          <th
            class=""
            data-key="4"
            data-label="Created at"
            scope="col"
          >
            Created at
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            Application Name
          </td>
          <td
            class=""
            data-key="1"
            data-label="State"
          >
            Some status
          </td>
          <td
            class=""
            data-key="2"
            data-label="Account"
          >
            Developer
          </td>
          <td
            class=""
            data-key="3"
            data-label="Plan"
          >
            Plan Name
          </td>
          <td
            class=""
            data-key="4"
            data-label="Created at"
          >
            1234567890
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
